sn_auditor/src/dag_db.rs

```diff
@@ -195,6 +213,86 @@ impl SpendDagDb {
         w_handle.merge(other)?;
         Ok(())
     }
+
+    /// Returns the current state of the beta program in JSON format
+    pub(crate) fn beta_program_json(&self) -> Result<String> {
+        let r_handle = self.beta_rewards.clone();
+        let beta_rewards = r_handle
+            .read()
+            .map_err(|e| eyre!("Failed to get beta rewards read lock: {e}"))?;
+        let json = serde_json::to_string_pretty(&*beta_rewards)?;
+        Ok(json)
+    }
+
+    /// Initialize beta program tracking and gathers current rewards from the DAG
+    pub(crate) async fn init_beta_program(&mut self, beta_participants: Vec<String>) -> Result<()> {
+        self.beta_participants = beta_participants
+            .iter()
+            .map(|h| (Hash::hash(h.as_bytes()), h.clone()))
+            .collect();
+        {
+            let w_handle = self.beta_rewards.clone();
+            let mut beta_rewards = w_handle
+                .write()
+                .map_err(|e| eyre!("Failed to get beta rewards write lock: {e}"))?;
+            *beta_rewards = beta_participants
+                .into_iter()
+                .map(|n| (n, BTreeSet::new()))
+                .collect();
+        }
+
+        // gather current rewards
+        self.gather_beta_rewards().await?;
+        Ok(())
+    }
+
+    // Gather beta rewards from the DAG
+    pub(crate) async fn gather_beta_rewards(&mut self) -> Result<()> {
+        info!("Gathering beta rewards...");
+
+        // get spends from current DAG
+        let r_handle = self.dag.clone();
+        let dag = r_handle
+            .read()
+            .map_err(|e| eyre!("Failed to get dag read lock for beta program: {e}"))?;
+        let all_spends = dag.all_spends();
+
+        // find spends with rewards
+        let mut rewards: BetaRewards = BTreeMap::new();
+        for spend in all_spends {
+            let user_name_hash = match spend.reason().get_sender_hash() {
+                Some(n) => n,
+                None => continue,
+            };
+            let addr = spend.address();
+            let amount = spend.spend.amount;
+            if let Some(user_name) = self.beta_participants.get(&user_name_hash) {
+                debug!("Got beta reward from {user_name} of {amount} at {addr:?}");
+                rewards
+                    .entry(user_name.to_owned())
+                    .or_default()
+                    .insert((addr, amount));
+            } else {
+                warn!(
```

Comment:
those transfers to credit the uploading clients, and those clients that paying the upload cost will be detected as `unknown participant` as well.
However they will be common.
So, maybe `warn` is too severe, `info` will be enough ?
or only warn when it is a `forward spend`, i.e. has only one input and one output (transfer or payment spend will have multiple outputs)